### ISSUE TITLE
cockroachdb: fails to create metadata table

### ISSUE THREAD

**Issue by dbist** [FerretDB/FerretDB#3252]
### FerretDB version

1.8.0

### Backend

CockroachDB 23.1.8

### Environment

- OS: the standard Docker image Linux ferretdb 5.15.49-linuxkit #1 SMP PREEMPT Tue Sep 13 07:51:32 UTC 2022 aarch64 GNU/Linux
- Deployment: Docker using ghcr.io/ferretdb/ferretdb:latest
- Deployment details: 4.19.0

### What did you do?

```javascript
test> db.test.insertOne({name: "Ada Lovelace", age: 205})
MongoServerError: [transactions.go:149 pgdb.(*Pool).InTransactionRetry] non-retriable error: [transactions.go:64 pgdb.(*Pool).InTransaction] [transactions.go:108 pgdb.(*Pool).InTransactionKeep] [error.go:372 commonerrors.CheckError] [insert.go:48 pgdb.InsertDocument] [collections.go:172 pgdb.CreateCollectionIfNotExists] [collections.go:129 pgdb.CreateCollection] [database_metadata.go:134 pgdb.(*metadataStorage).store] [indexes.go:247 pgdb.createPgIndexIfNotExists] ERROR: index element (_jsonb->'_id') of type jsonb is not indexable in a non-inverted index (SQLSTATE 42P16)
```

### What did you expect to see?

metadata table should be created.

### What did you see instead?

```bash
2023-08-25T14:31:10.677Z        INFO    pgdb    tracelog/tracelog.go:335        Query   {"time": "2.153458ms", "commandTag": "ROLLBACK", "pid": 2917020, "sql": "rollback", "args": []}
2023-08-25T14:31:10.677Z        INFO    pgdb    tracelog/tracelog.go:335        Query   {"args": [], "time": "212.791µs", "commandTag": "BEGIN", "pid": 2917020, "sql": "begin"}
2023-08-25T14:31:10.699Z        INFO    pgdb    tracelog/tracelog.go:335        Query   {"commandTag": "SELECT 1", "pid": 2917020, "sql": "SELECT EXISTS ( SELECT 1 FROM information_schema.columns WHERE table_schema = $1 AND table_name = $2 )", "args": ["test","_ferretdb_database_metadata"], "time": "22.165083ms"}
2023-08-25T14:31:10.702Z        INFO    pgdb    tracelog/tracelog.go:335        Query   {"pid": 2917020, "sql": "CREATE SCHEMA IF NOT EXISTS \"test\"", "args": [], "time": "2.460333ms", "commandTag": "CREATE SCHEMA"}
2023-08-25T14:31:10.705Z        INFO    pgdb    tracelog/tracelog.go:335        Query   {"pid": 2917020, "sql": "CREATE TABLE IF NOT EXISTS \"test\".\"_ferretdb_database_metadata\" (_jsonb jsonb)", "args": [], "time": "3.046834ms", "commandTag": "CREATE TABLE"}
2023-08-25T14:31:10.706Z        ERROR   pgdb    tracelog/tracelog.go:335        Query   {"sql": "CREATE UNIQUE INDEX IF NOT EXISTS \"_ferretdb_database_metadata_id_idx\" ON \"test\".\"_ferretdb_database_metadata\" (((_jsonb->'_id')) ASC)", "args": [], "err": "ERROR: index element (_jsonb->'_id') of type jsonb is not indexable in a non-inverted index (SQLSTATE 42P16)", "time": "933.834µs", "pid": 2917020}
github.com/jackc/pgx/v5/tracelog.(*TraceLog).log
        /cache/gomodcache/github.com/jackc/pgx/v5@v5.4.2/tracelog/tracelog.go:335
github.com/jackc/pgx/v5/tracelog.(*TraceLog).TraceQueryEnd
        /cache/gomodcache/github.com/jackc/pgx/v5@v5.4.2/tracelog/tracelog.go:160
github.com/FerretDB/FerretDB/internal/handlers/pg/pgdb.(*multiQueryTracer).TraceQueryEnd
        /src/internal/handlers/pg/pgdb/trace.go:74
github.com/jackc/pgx/v5.(*Conn).Exec
        /cache/gomodcache/github.com/jackc/pgx/v5@v5.4.2/conn.go:417
github.com/jackc/pgx/v5.(*dbTx).Exec
        /cache/gomodcache/github.com/jackc/pgx/v5@v5.4.2/tx.go:222
github.com/jackc/pgx/v5/pgxpool.(*Tx).Exec
        /cache/gomodcache/github.com/jackc/pgx/v5@v5.4.2/pgxpool/tx.go:69
github.com/FerretDB/FerretDB/internal/handlers/pg/pgdb.createPgIndexIfNotExists
        /src/internal/handlers/pg/pgdb/indexes.go:233
github.com/FerretDB/FerretDB/internal/handlers/pg/pgdb.(*metadataStorage).store
        /src/internal/handlers/pg/pgdb/database_metadata.go:133
github.com/FerretDB/FerretDB/internal/handlers/pg/pgdb.CreateCollection
        /src/internal/handlers/pg/pgdb/collections.go:127
github.com/FerretDB/FerretDB/internal/handlers/pg/pgdb.CreateCollectionIfNotExists
        /src/internal/handlers/pg/pgdb/collections.go:164
github.com/FerretDB/FerretDB/internal/handlers/pg/pgdb.InsertDocument
        /src/internal/handlers/pg/pgdb/insert.go:47
github.com/FerretDB/FerretDB/internal/handlers/pg.insertDocument
        /src/internal/handlers/pg/msg_insert.go:133
github.com/FerretDB/FerretDB/internal/handlers/pg.insertDocumentSeparately.func1
        /src/internal/handlers/pg/msg_insert.go:162
github.com/FerretDB/FerretDB/internal/handlers/pg/pgdb.(*Pool).InTransaction.func1
        /src/internal/handlers/pg/pgdb/transactions.go:61
github.com/FerretDB/FerretDB/internal/handlers/pg/pgdb.(*Pool).InTransactionKeep
        /src/internal/handlers/pg/pgdb/transactions.go:107
github.com/FerretDB/FerretDB/internal/handlers/pg/pgdb.(*Pool).InTransaction
        /src/internal/handlers/pg/pgdb/transactions.go:59
github.com/FerretDB/FerretDB/internal/handlers/pg/pgdb.(*Pool).InTransactionRetry
        /src/internal/handlers/pg/pgdb/transactions.go:128
github.com/FerretDB/FerretDB/internal/handlers/pg.insertDocumentSeparately
        /src/internal/handlers/pg/msg_insert.go:161
github.com/FerretDB/FerretDB/internal/handlers/pg.insertMany
        /src/internal/handlers/pg/msg_insert.go:102
github.com/FerretDB/FerretDB/internal/handlers/pg.(*Handler).MsgInsert
        /src/internal/handlers/pg/msg_insert.go:55
github.com/FerretDB/FerretDB/internal/clientconn.(*conn).handleOpMsg
        /src/internal/clientconn/conn.go:563
github.com/FerretDB/FerretDB/internal/clientconn.(*conn).route
        /src/internal/clientconn/conn.go:426
github.com/FerretDB/FerretDB/internal/clientconn.(*conn).run
        /src/internal/clientconn/conn.go:310
github.com/FerretDB/FerretDB/internal/clientconn.acceptLoop.func1
        /src/internal/clientconn/listener.go:321
2023-08-25T14:31:10.708Z        INFO    pgdb    tracelog/tracelog.go:335        Query   {"args": [], "time": "1.891667ms", "commandTag": "ROLLBACK", "pid": 2917020, "sql": "rollback"}
2023-08-25T14:31:57.288Z        INFO    listener        clientconn/listener.go:324      Connection stopped      {"conn": "172.18.0.11:34364 -> 172.18.0.8:27017"}
2023-08-25T14:31:57.290Z        INFO    listener        clientconn/listener.go:324      Connection stopped      {"conn": "172.18.0.11:34366 -> 172.18.0.8:27017"}
2023-08-25T14:31:57.291Z        INFO    listener        clientconn/listener.go:324      Connection stopped      {"conn": "172.18.0.11:34390 -> 172.18.0.8:27017"}
2023-08-25T14:31:57.292Z        INFO    listener        clientconn/listener.go:324      Connection stopped      {"conn": "172.18.0.11:34376 -> 172.18.0.8:27017"}
```

```sql
CREATE UNIQUE INDEX IF NOT EXISTS _ferretdb_database_metadata_id_idx ON test._ferretdb_database_metadata ((( _jsonb->'_id')) ASC); 

ERROR: index element (_jsonb->'_id') of type jsonb is not indexable


CREATE UNIQUE INDEX IF NOT EXISTS _ferretdb_database_metadata_id_idx ON test._ferretdb_database_metadata USING GIN ((( _jsonb->'_id')) ASC); 

ERROR: inverted indexes can't be unique
```

**Comment by AlekSi** [comment]
The real fix would be in #226, but for #3228 we could change that index to be a GIN non-unique index for the CockroachDB backend.

**Comment by dbist** [comment]
I tested again with 1.9.0 and the issue seems to persist. I was trying to workaround using the following

```sql
CREATE TABLE IF NOT EXISTS test._ferretdb_database_metadata (
    _id STRING PRIMARY KEY AS (_jsonb->>'_id') STORED,
    _jsonb jsonb
);
ALTER TABLE test._ferretdb_database_metadata ALTER COLUMN _id DROP STORED;
```

Seems, FerretDB is trying to create the index irrelevant of the UNIQUE constraint existence

```sql
CREATE TABLE IF NOT EXISTS test._ferretdb_database_metadata (
    _id STRING PRIMARY KEY AS (_jsonb->>'_id') STORED,
    _jsonb jsonb
);
ALTER TABLE test._ferretdb_database_metadata ALTER COLUMN _id DROP STORED;

CREATE UNIQUE INDEX _ferretdb_database_metadata_id_idx ON test._ferretdb_database_metadata (_id);
```

```bash
MongoServerError: [transactions.go:149 pgdb.(*Pool).InTransactionRetry] non-retriable error: [transactions.go:64 pgdb.(*Pool).InTransaction] [transactions.go:108 pgdb.(*Pool).InTransactionKeep] [msg_insert.go:156 pg.insertDocument] [insert.go:48 pgdb.InsertDocument] [collections.go:172 pgdb.CreateCollectionIfNotExists] [collections.go:129 pgdb.CreateCollection] [database_metadata.go:134 pgdb.(*metadataStorage).store] [indexes.go:247 pgdb.createPgIndexIfNotExists] ERROR: index element (_jsonb->'_id') of type jsonb is not indexable in a non-inverted index (SQLSTATE 42P16)
```

is there a way to add a check that a unique index already exists? 

https://github.com/FerretDB/FerretDB/blob/main/internal/handlers/pg/pgdb/indexes.go#L230-L247

I know you're relying on PG to check the index existence but can we add a Cockroach condition?

**Comment by dbist** [comment]
My team is also reporting that in the next release, slated for the end of the year, the following statement should work

```sql
CREATE UNIQUE INDEX IF NOT EXISTS _ferretdb_database_metadata_id_idx ON test._ferretdb_database_metadata ((( _jsonb->'_id')) ASC); 
```

so perhaps this is a no-op issue and we just wait?

**Comment by dbist** [comment]
confirmed, in 23.2, FerretDB works as advertised

```bash
test> db.test.insertOne({name: "Ada Lovelace", age: 205})
{
  acknowledged: true,
  insertedId: ObjectId("64ed15ccac5241367eadac84")
}
test> 
```

```bash
2023-08-28T21:46:52.157Z        INFO    pgdb    tracelog/tracelog.go:335        Query   {"args": ["7b222473223a7b2270223a7b225f6964223a7b2274223a22737472696e67227d2c227461626c65223a7b2274223a22737472696e67227d2c22696e6465786573 (truncated 386 bytes)","227465737422"], "time": "1.39775ms", "commandTag": "UPDATE 1", "pid": 3810834, "sql": "UPDATE \"test\".\"_ferretdb_database_metadata\" SET _jsonb = $1 WHERE _jsonb->'_id' = $2"}
2023-08-28T21:46:52.159Z        INFO    pgdb    tracelog/tracelog.go:335        Query   {"sql": "CREATE UNIQUE INDEX IF NOT EXISTS \"test__id__74791483_idx\" ON \"test\".\"test_afd071e5\" (((_jsonb->'_id')) ASC)", "args": [], "time": "1.02275ms", "commandTag": "CREATE INDEX", "pid": 3810834}
2023-08-28T21:46:52.172Z        INFO    pgdb    tracelog/tracelog.go:335        Query   {"pid": 3810834, "sql": "SELECT EXISTS ( SELECT 1 FROM information_schema.columns WHERE table_schema = $1 AND table_name = $2 )", "args": ["test","_ferretdb_database_metadata"], "time": "13.8365ms", "commandTag": "SELECT 1"}
2023-08-28T21:46:52.174Z        INFO    pgdb    tracelog/tracelog.go:335        Query   {"sql": "SELECT _jsonb  FROM \"test\".\"_ferretdb_database_metadata\" WHERE _jsonb->$1 @> $2", "args": ["_id","\"test\""], "time": "1.128958ms", "commandTag": "SELECT 1", "pid": 3810834}
2023-08-28T21:46:52.177Z        INFO    pgdb    tracelog/tracelog.go:335        Query   {"args": ["7b222473223a7b2270223a7b225f6964223a7b2274223a226f626a6563744964227d2c226e616d65223a7b2274223a22737472696e67227d2c22616765223a7b (truncated 104 bytes)"], "time": "2.841958ms", "commandTag": "INSERT 0 1", "pid": 3810834, "sql": "INSERT INTO \"test\".\"test_afd071e5\" (_jsonb) VALUES ($1)"}
2023-08-28T21:46:52.265Z        INFO    pgdb    tracelog/tracelog.go:335        Query   {"sql": "commit", "args": [], "time": "87.837084ms", "commandTag": "COMMIT", "pid": 3810834}
```

```sql
root@lb:26000/ferretdb> select * from test.test_afd071e5;                                                                             
                                                                                             _jsonb
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  {"$s": {"$k": ["_id", "name", "age"], "p": {"_id": {"t": "objectId"}, "age": {"t": "int"}, "name": {"t": "string"}}}, "_id": "64ed15ccac5241367eadac84", "age": 205, "name": "Ada Lovelace"}
(1 row)
```

so the question is, do we want to leave this integration as broken until the new release is out?

**Comment by AlekSi** [comment]
I made a quick hack to disable the creation of unique indexes: #3272. Links to Docker images will be there in about an hour: https://github.com/FerretDB/FerretDB/actions/runs/6012212211?pr=3272

Looking into https://www.cockroachlabs.com/docs/stable/jsonb, I think there might be more problems with the current version. But please check that quick hack – it should give us more info to think about.

**Comment by dbist** [comment]
It didn't work but for a different reason altogether. 

```bash
test> db.test.insertOne({name: "Ada Lovelace", age: 205})
MongoServerError: [msg_insert.go:37 pg.(*Handler).MsgInsert] [pg.go:140 pg.(*Handler).DBPool] [pool.go:109 pgdb.NewPool] [pool.go:176 pgdb.(*Pool).checkConnection] failed to connect to `host=lb user=root database=ferretdb`: server error (ERROR: invalid value for parameter "search_path": "" (SQLSTATE 22P02))
test> 
```

I tried to add `SELECT pg_catalog.set_config('search_path', 'public', false);` in a schema migration prior to ferretdb starting but it didn't have an effect on the outcome. Neither did passing the search_path in the FerretDB url. I believe I have to run FerretDB in non-debug build to overcome it but I'm not sure if it's hard-coded in your custom build of the image? I'm happy to continue down this path but is it worth having a one-off build for Cockroach, I already confirmed it works in the next release with the UNIQUE constraint? Also, should I open an issue for a `search_path` in debug builds? I know that will trip users up. https://docs.ferretdb.io/configuration/flags/

**Comment by AlekSi** [comment]
The next build will appear soon here: https://github.com/FerretDB/FerretDB/actions/runs/6015908634/job/16318769010?pr=3272. It disables setting `search_path` to an empty string that is useful during FerretDB development with PostgreSQL: that way we ensure that we always add the schema.

> I'm not sure if it's hard-coded in your custom build of the image

Yeah, that's a "feature" of debug builds: https://github.com/FerretDB/FerretDB/blob/2e847f06d28102964ad99a6a485b674b150d50b4/internal/handlers/pg/pgdb/pool.go#L142-L146

> should I open an issue for a search_path in debug builds

Not yet. We will probably remove that code soon.

> but is it worth having a one-off build for Cockroach, I already confirmed it works in the next release with the UNIQUE constraint

Basically, we are trying to understand how much time it would take us to have a proper fix for the current version of CockroachDB.

**Comment by dbist** [comment]
we're back to step 1 with the following error:

```bash
MongoServerError: [transactions.go:149 pgdb.(*Pool).InTransactionRetry] non-retriable error: [transactions.go:64 pgdb.(*Pool).InTransaction] [transactions.go:108 pgdb.(*Pool).InTransactionKeep] [msg_insert.go:156 pg.insertDocument] [insert.go:48 pgdb.InsertDocument] [collections.go:172 pgdb.CreateCollectionIfNotExists] [collections.go:129 pgdb.CreateCollection] [database_metadata.go:134 pgdb.(*metadataStorage).store] [indexes.go:247 pgdb.createPgIndexIfNotExists] ERROR: index element (_jsonb->'_id') of type jsonb is not indexable in a non-inverted index (SQLSTATE 42P16)
```

**Comment by AlekSi** [comment]
It should work now with the current version of FerretDB and CockroachDB v23.2.

**Comment by AlekSi** [comment]
It is out of beta
